rotkehlchen/chain/ethereum/modules/makerdao/sai/decoder.py

```diff
@@ -0,0 +1,385 @@
+                _event.asset == self.sai and
+                _event.balance.amount == amount_paid
+            ):
+                already_decoded = True
+                return None, []
+
+        if already_decoded is False:
+            event = HistoryBaseEntry(
+                event_identifier=transaction.tx_hash,
+                timestamp=ts_sec_to_ms(transaction.timestamp),
+                sequence_index=self.base_tools.get_next_sequence_counter(),
+                location=Location.BLOCKCHAIN,
+                asset=self.sai,
+                event_type=HistoryEventType.SPEND,
+                event_subtype=HistoryEventSubType.PAYBACK_DEBT,
+                balance=Balance(amount=amount_paid),
+                location_label=depositor,
+                counterparty=CPT_SAI,
+                notes=f'Repay {amount_paid} {self.sai.symbol} to CDP {cdp_id}',
+            )
+            return event, []
+
+        return None, []
+
+    def _decode_makerdao_liquidate_event(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: List[HistoryBaseEntry],
+            all_logs: List[EvmTxReceiptLog],
+            action_items: Optional[List[ActionItem]],  # pylint: disable=unused-argument
+    ) -> Tuple[Optional[HistoryBaseEntry], List[ActionItem]]:
+        liquidator = hex_or_bytes_to_address(tx_log.topics[1])
+        cdp_id = hex_or_bytes_to_int(tx_log.topics[2])
+
+        already_decoded = False
+        if already_decoded is False:
+            for event in decoded_events:
+                if (
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.GENERATE_DEBT and
+                    event.asset == self.peth and
+                    event.counterparty == CPT_SAI
+                ):
+                    already_decoded = True
+                    return None, []
+
+        # check for the transfer event of the liquidation to Maker SaiTap contract
+        for _tx_log in all_logs:
+            if len(_tx_log.topics) < 2:
+                continue
+
+            if hex_or_bytes_to_address(_tx_log.topics[1]) == MAKERDAO_SAITUB_CONTRACT and hex_or_bytes_to_address(_tx_log.topics[2]) == MAKERDAO_SAITAP_CONTRACT:  # noqa: E501
+                amount_raw = hex_or_bytes_to_int(_tx_log.data[:32])
+                amount = asset_normalized_value(amount=amount_raw, asset=self.weth)
+
+                event = HistoryBaseEntry(
+                    event_identifier=transaction.tx_hash,
+                    timestamp=ts_sec_to_ms(transaction.timestamp),
+                    sequence_index=self.base_tools.get_next_sequence_counter(),
+                    location=Location.BLOCKCHAIN,
+                    asset=self.peth,
+                    balance=Balance(amount=amount),
+                    event_type=HistoryEventType.SPEND,
+                    event_subtype=HistoryEventSubType.GENERATE_DEBT,
```

Comment:
Question. Words `generate debt` and `liquidate` sound kinda opposite to me. Or is everything correct here?